File: src/leetcode/2026/05/Solution942.java
public class Solution942 {
    public int[] diStringMatch(String s) {
        int n = s.length();
        int low = 0;
        int high = n;
        int[] result = new int[n + 1];

        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == 'I') {
                result[i] = low++;
            } else {
                result[i] = high--;
            }
        }

        result[n] = low; // At this point, low and high should be equal

        return result;
    }

    public static void main(String[] args) {
        Solution942 solution = new Solution942();

        String s1 = "IDID";
        int[] result1 = solution.diStringMatch(s1);
        System.out.print("Output for s1: ");
        for (int num : result1) {
            System.out.print(num + " ");
        }
        System.out.println(); // Output: [0, 4, 1, 3, 2]

        String s2 = "III";
        int[] result2 = solution.diStringMatch(s2);
        System.out.print("Output for s2: ");
        for (int num : result2) {
            System.out.print(num + " ");
        }
        System.out.println(); // Output: [0, 1, 2, 3]

        String s3 = "DDI";
        int[] result3 = solution.diStringMatch(s3);
        System.out.print("Output for s3: ");
        for (int num : result3) {
            System.out.print(num + " ");
        }
        System.out.println(); // Output: [3, 2, 0, 1]
    }
}

File: src/leetcode/2026/05/Solution942.cpp
#include <vector>
#include <string>
#include <iostream>
using namespace std;

class Solution942 {
public:
    vector<int> diStringMatch(string s) {
        int n = s.size();
        int low = 0, high = n;
        vector<int> result(n + 1);

        for (int i = 0; i < n; i++) {
            if (s[i] == 'I') {
                result[i] = low++;
            } else {
                result[i] = high--;
            }
        }

        result[n] = low; // or result[n] = high; since low == high at this point
        return result;
    }
};

int main() {
    Solution942 solution;

    string s1 = "IDID";
    string s2 = "III";
    string s3 = "DDI";

    vector<int> result1 = solution.diStringMatch(s1);
    vector<int> result2 = solution.diStringMatch(s2);
    vector<int> result3 = solution.diStringMatch(s3);

    for (int num : result1) {
        cout << num << " ";
    }
    cout << endl;

    for (int num : result2) {
        cout << num << " ";
    }
    cout << endl;

    for (int num : result3) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}